### ISSUE TITLE
Device disconnected message

### ISSUE THREAD

**Issue by ItsLemmy** [t3knomanzer/maxmix-software#128]
**Is your feature request related to a problem? Please describe.**
If MaxMix software is not launched (or has crashed) the device continue to behave normally, but it is not doing anything

**Describe the solution you'd like**
When not asleep, the device's display could show a disconnected message.

**Comment by t3knomanzer** [comment]
How would you go about it?
Implement message confirmation in the app? That would only detect they the application has crashed after the next message is sent once the app has crashed.  
Unless a connection status message with confirmation is sent regularly i guess.

**Comment by ItsLemmy** [comment]
I guess a kind of ping message sent every second from the device to the app would do the job.
We have a bunch of USB bandwidth available and it would be very reliable.

**Comment by t3knomanzer** [comment]
We do need to implement message acknowledgment anyways so maybe take the chance to do both? 

The reason for acknowledging reception is that messages can be sent too quickly by the app to the device and they get lost.  
Currently there is temporary hack, a delay of 50ms in the application to prevent it from happening. The side effect however is that we’re limiting at 20 messages/second which is super slow.   You can see this if you open the volume mixer and drag the volume slider of an application. 
Implementing acknowledge would eliminate the artificial wait. 

And once the system is in place, as you said, we can just send a ping message and wait for acknowledge/timeout. 

What do you think?

**Comment by ItsLemmy** [comment]
I haven't checked much the app code yet, will do soon. I assumed there was already some acknowledgment in place.
I guess this is the perfect occasion to do both and make the entire communication more robust.

**Comment by t3knomanzer** [comment]
@quadbyte  Nah there isn’t. It’s been long overdue. 
If you want to take a stab at it would be great. I started implementing it in a branch but wasn’t happy with it at all.

**Comment by ItsLemmy** [comment]
Sure, will look into it asap.

**Comment by SchemingWeasels** [comment]
@quadbyte are you already working on this? I have _some_ time this weekend. This should be a quick addition.

**Comment by ItsLemmy** [comment]
@SchemingWeasels Haven't started yet, feel free  :)

**Comment by SchemingWeasels** [comment]
#142 is one-way acknowledgment? Device-> Desktop only?

**Comment by t3knomanzer** [comment]
Yep.

**Comment by SchemingWeasels** [comment]
Aite. Have a bit more work to do then.

**Comment by SchemingWeasels** [comment]
Ok, so I have a few ideas.
**1. Heartbeat - Keep-alive packet**
**Pros**: Device knows instantly when app crashes.
**Cons**: Additional unnecessary serial messages, shouldn't be an issue for the computer, but sounds a bit sloppy for the Arduino Nano.
**Implementation idea:** Nano sends a heartbeat request on a timer, host just responds. Timeout = reset arduino.  This will put the nano back to waiting for HS and display the logo. Using the existing Comns service, the revision number will quickly overflow, would be this ok? As long they both overflow the same way it should be fine right?
Also ,the current Send service is expecting an ack reply as well, the nano also sends an ack packet. So for example, < Nano sends HBReq >, < Host sends Keep-Alive >, < Nano sends Ack >

**2. 2-way receive acknowledgment**
**Pros**: Less overhead, fewer changes to current code.
**Cons**: Device only knows the app crashed when trying to send a command.
**Implementation idea:** Nano -> computer messages will be replied with an ack packet, but since the current Send service is expecting an ack reply as well, the nano also sends an ack packet. So for example, < Nano sends UpdateVol >, < Host sends Ack >, < Nano sends Ack >

**3. Heartbeat on input**
So basically number 1, but instead of running on a timer, it checks on user input. Is it essential to know that the host crashed immediately? After all, if I reset the arduino it will show the logo and go to sleep anyway. So you'll only find out if it crashed once you wake the device. Unless we don't reset it, instead we display a disconnected message. But I'd rather reset it so that it's ready to reconnect once you relaunch the app.

Or I can rewrite the Sending service to either expect an ack or not, then we dont need the final Nano -> Host ack

**Comment by ItsLemmy** [comment]
Number 2 => 2 Way ACK sounds good to me.  A simple check on the nano side to avoid sending ACK when it receives an ACK from the host should do the job.

**Comment by t3knomanzer** [comment]
Sounds good!

**Comment by ItsLemmy** [comment]
That being said, I'm working on the Discovery + Communication services to make it more robust, so you may want to wait a bit for me to commit my changes (hopefully tomorrow) before adding more complexity in there, as my changes will refactor many things.

**Comment by SchemingWeasels** [comment]
Hey @quadbyte. I noticed SendItemVolumeCommand in the fw is sending a packageRevision to the host, but it doesn't seem to be used or processed anywhere. Was this left in for future ack handling? Or did I miss some code somewhere?

**Comment by ItsLemmy** [comment]
Hi @SchemingWeasels 
Yes, it was left for a future 2way-ack handling and is unused for now.
In the Arduino code you may want to rename packageRevision to messageRevision (that's now how it's called on the host side, see upcoming PR : https://github.com/t3knomanzer/maxmix-software/pull/150)
Thanks !

**Comment by SchemingWeasels** [comment]
https://github.com/t3knomanzer/maxmix-software/blob/e2ea7162ff671ea8001232aaed9953db450c848b/Desktop/Application/MaxMix/Services/Communication/Serial/CobsSerializationService%20.cs#L207-L223

If the device sends a volume update, the message revision is lost because only an ack message retains that byte. 
Do I add a revision var to every msg type? Or perhaps pass it as the first payload byte. 
I want to retrieve the revision so that I can send the received revision back to the arduino for error checking, like how the host handles acks now. 
Or perhaps we simplify stuff and don't expect a revision number back?

**Comment by t3knomanzer** [comment]
I’ll leave that to @quadbyte since he did the initial implementation.

**Comment by SchemingWeasels** [comment]
I think we add revisions to every message since we want to support changing output devices and mics, may as well code with supporting future changes in mind?

**Comment by ItsLemmy** [comment]
I would simply add messageRevision to every messages.  The revision is already there in every messages, both ways, just not used on message originating from the device when they are not MessageAcknowledgment.

**Comment by ItsLemmy** [comment]
At this point, to avoid slowing the serial comms I suggest we implement a simple heartbeat every 3 seconds from the device to the app. If no answer, reset the device. So people can notice if the app crash.
I can take care of it.
